core/access/darc/types/permission.go

```diff
@@ -0,0 +1,161 @@
+package types
+
+import (
+	"go.dedis.ch/dela/core/access"
+	"go.dedis.ch/dela/crypto/common"
+	"go.dedis.ch/dela/serde"
+	"go.dedis.ch/dela/serde/registry"
+	"golang.org/x/xerrors"
+)
+
+var permFormats = registry.NewSimpleRegistry()
+
+// RegisterPermissionFormat registers the engine for the provided format.
+func RegisterPermissionFormat(c serde.Format, f serde.FormatEngine) {
+	permFormats.Register(c, f)
+}
+
+// DisjunctivePermission is a permission implementation that is using the
+// Disjunctive Normal Form to represent the groups of identities allowed for a
+// given rule.
+//
+// - implements darc.Permission
+type DisjunctivePermission struct {
+	rules map[string]*Expression
+}
+
+// PermissionOption is the option type to create an access control.
+type PermissionOption func(*DisjunctivePermission)
+
+// WithRule is an option to grant a given group access to a rule.
+func WithRule(rule string, group ...access.Identity) PermissionOption {
+	return func(perm *DisjunctivePermission) {
+		perm.Evolve(rule, true, group...)
+	}
+}
+
+// WithExpression is an option to set a rule from its expression.
+func WithExpression(rule string, expr *Expression) PermissionOption {
+	return func(perm *DisjunctivePermission) {
+		perm.rules[rule] = expr
+	}
+}
+
+// NewPermission returns a new empty instance of an access control.
+func NewPermission(opts ...PermissionOption) *DisjunctivePermission {
+	a := &DisjunctivePermission{
+		rules: make(map[string]*Expression),
+	}
+
+	for _, opt := range opts {
+		opt(a)
+	}
+
+	return a
+}
+
+// GetRules returns a map of the expressions.
+func (perm *DisjunctivePermission) GetRules() map[string]*Expression {
+	rules := make(map[string]*Expression)
+
+	for rule, expr := range perm.rules {
+		rules[rule] = expr
+	}
+
+	return rules
+}
+
+// Evolve implements darc.Permission. It grants or remove the access to a group
+// to a given rule.
+func (perm *DisjunctivePermission) Evolve(rule string, grant bool, group ...access.Identity) {
+	expr, ok := perm.rules[rule]
+	if !ok {
+		if !grant {
+			return
+		}
+
+		expr = NewExpression()
+	}
+
+	expr.Evolve(grant, group)
+
+	if len(expr.matches) == 0 {
```

Comment:
this stuff is hard to get when reading the code. Why would it delete the rule in case the expression evolve doesn't have any match ?

Reply:
I guess the evolve should be split into Allow and Deny as this is just a simple cleaning procedure after removing a group.

Reply:
well there was a bug with the removal so it is indeed better when split.